### PR TITLE
support for inlining haskell functions to logic

### PR DESCRIPTION
**Pull request by nikivazou**
see `tests/pos/inline.hs`

### CODE REVIEW

**Comment by ranjitjhala** [comment]
nice!



On Thu, Feb 5, 2015 at 1:11 PM, Niki Vazou <notifications@github.com> wrote:

> see tests/pos/inline.hs
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/ucsd-progsys/liquidhaskell/pull/325
> Commit Summary
>
>    - support for inlining haskell functions to logic
>
> File Changes
>
>    - *M* src/Language/Haskell/Liquid/Bare/Env.hs
>    <https://github.com/ucsd-progsys/liquidhaskell/pull/325/files#diff-0>
>    (17)
>    - *M* src/Language/Haskell/Liquid/Bare/GhcSpec.hs
>    <https://github.com/ucsd-progsys/liquidhaskell/pull/325/files#diff-1>
>    (38)
>    - *M* src/Language/Haskell/Liquid/Bare/Measure.hs
>    <https://github.com/ucsd-progsys/liquidhaskell/pull/325/files#diff-2>
>    (43)
>    - *M* src/Language/Haskell/Liquid/Bare/RefToLogic.hs
>    <https://github.com/ucsd-progsys/liquidhaskell/pull/325/files#diff-3>
>    (70)
>    - *M* src/Language/Haskell/Liquid/Bare/Spec.hs
>    <https://github.com/ucsd-progsys/liquidhaskell/pull/325/files#diff-4>
>    (4)
>    - *M* src/Language/Haskell/Liquid/Constraint/Generate.hs
>    <https://github.com/ucsd-progsys/liquidhaskell/pull/325/files#diff-5>
>    (6)
>    - *M* src/Language/Haskell/Liquid/CoreToLogic.hs
>    <https://github.com/ucsd-progsys/liquidhaskell/pull/325/files#diff-6>
>    (29)
>    - *M* src/Language/Haskell/Liquid/Measure.hs
>    <https://github.com/ucsd-progsys/liquidhaskell/pull/325/files#diff-7>
>    (3)
>    - *M* src/Language/Haskell/Liquid/Parse.hs
>    <https://github.com/ucsd-progsys/liquidhaskell/pull/325/files#diff-8>
>    (9)
>    - *A* tests/pos/inline.hs
>    <https://github.com/ucsd-progsys/liquidhaskell/pull/325/files#diff-9>
>    (54)
>
> Patch Links:
>
>    - https://github.com/ucsd-progsys/liquidhaskell/pull/325.patch
>    - https://github.com/ucsd-progsys/liquidhaskell/pull/325.diff
>
> --
> Reply to this email directly or view it on GitHub
> <https://github.com/ucsd-progsys/liquidhaskell/pull/325>.
>



-- 
Ranjit.

File: src/Language/Haskell/Liquid/Bare/RefToLogic.hs [hunk 1]
```diff
@@ -52,7 +62,8 @@ instance Transformable Pred where
 	tx s m (PNot p)        = PNot (tx s m p)
 	tx s m (PImp p1 p2)    = PImp (tx s m p1) (tx s m p2)
 	tx s m (PIff p1 p2)    = PIff (tx s m p1) (tx s m p2)
-	tx s m (PBexp e)       = PBexp (tx s m e)
+	tx s m (PBexp (EApp f es)) = txPApp (s, m) f es
+	tx s m (PBexp e)       = PBexp (tx s m $ traceShow "GOTO EXPR" e)
```
Comment (spinda) [inline src/Language/Haskell/Liquid/Bare/RefToLogic.hs]: DId you mean to leave these traces in?

File: src/Language/Haskell/Liquid/CoreToLogic.hs [hunk 2]
```diff
@@ -145,7 +146,24 @@ coreToDef x _ e = go $ inline_preds $ simplify e
 
     inline_preds = inline (eqType boolTy . varType)
 
+coreToFun :: LocSymbol -> Var -> C.CoreExpr ->  LogicM ([Symbol], Either Pred Expr)
+coreToFun _ v e = go [] $ inline_preds $ simplify $ traceShow "coreToFun" e
```
Comment (spinda) [inline src/Language/Haskell/Liquid/CoreToLogic.hs]: (more traces here as well)

**Comment by spinda** [comment]
Awesome! Can predicate and expression aliases be phased out eventually with this in?

(Also, the build is failing because Travis [messed up](https://travis-ci.org/ucsd-progsys/liquidhaskell/jobs/49674984) on one test set; needs restarting.)

**Comment by spinda** [comment]
I noticed that bad inlines don't get location information on their error messages.

```haskell
module Fixme where

z :: Int
z = 1

{-@ inline eqq @-}
eqq :: Int -> Int -> Bool
eqq x y = x > z
```

```
 Fixpoint.Types.dummyLoc:0:0: Error: Bad Type Specification
 Fixme.eqq :: x1:Int -> x2:Int -> Bool | True
     Sort Error in Refinement: {VV : GHC.Types.Bool | [Prop VV <=> x1 > Fixme.z#rou]}
     Unbound Symbol Fixme.z#rou
 Perhaps you meant: Fixme.eqq#rpv
```

I hacked up a quick patch that would add it in. The location information is present, but it's discarded in `makeHIMeas`.

```diff
diff --git a/src/Language/Haskell/Liquid/Bare/Measure.hs b/src/Language/Haskell/Liquid/Bare/Measure.hs
index b1d02bf..ffb68ed 100644
--- a/src/Language/Haskell/Liquid/Bare/Measure.hs
+++ b/src/Language/Haskell/Liquid/Bare/Measure.hs
@@ -123,8 +123,8 @@ makeMeasureDefinition lmap cbs x
 simplesymbol = symbol . getName
 
 
-strengthenHaskellMeasures :: S.HashSet Var -> [(Var, Located SpecType)]
-strengthenHaskellMeasures hmeas = (\v -> (v, dummyLoc $ strengthenResult v)) <$> (S.toList hmeas)
+strengthenHaskellMeasures :: S.HashSet (Located Var) -> [(Var, Located SpecType)]
+strengthenHaskellMeasures hmeas = (\v -> (val v, fmap strengthenResult v)) <$> (S.toList hmeas)
 
 makeMeasureSelectors :: (DataCon, Located DataConP) -> [Measure SpecType DataCon]
 makeMeasureSelectors (dc, (Loc loc (DataConP _ vs _ _ _ xts r))) = go <$> zip (reverse xts) [1..]
diff --git a/src/Language/Haskell/Liquid/Bare/Spec.hs b/src/Language/Haskell/Liquid/Bare/Spec.hs
index c79980b..5b4c4bc 100644
--- a/src/Language/Haskell/Liquid/Bare/Spec.hs
+++ b/src/Language/Haskell/Liquid/Bare/Spec.hs
@@ -82,8 +82,8 @@ makeQualifiers (mod,spec) = inModule mod mkQuals
 makeHints vs spec = varSymbols id vs $ Ms.decr spec
 makeLVar  vs spec = fmap fst <$> (varSymbols id vs $ [(v, ()) | v <- Ms.lvars spec])
 makeLazy  vs spec = fmap fst <$> (varSymbols id vs $ [(v, ()) | v <- S.toList $ Ms.lazy spec])
-makeHIMeas vs spec = fmap fst <$> (varSymbols id vs $ [(v, loc v) | v <- (S.toList $ Ms.hmeas spec) ++ (S.toList $ Ms.inlines spec)])
 makeTExpr vs spec = varSymbols id vs $ Ms.termexprs spec
+makeHIMeas vs spec = fmap (uncurry $ flip Loc) <$> (varSymbols id vs $ [(v, loc v) | v <- (S.toList $ Ms.hmeas spec) ++ (S.toList $ Ms.inlines spec)])
 
 varSymbols :: ([Var] -> [Var]) -> [Var] -> [(LocSymbol, a)] -> BareM [(Var, a)]
 varSymbols f vs  = concatMapM go
```

Result:

```
 tests/pos/inline.hs:6:12: Error: Bad Type Specification
 Fixme.eqq :: x1:Int -> x2:Int -> Bool | True
     Sort Error in Refinement: {VV : GHC.Types.Bool | [Prop VV <=> x1 > Fixme.z#rou]}
     Unbound Symbol Fixme.z#rou
 Perhaps you meant: Fixme.eqq#rpv
```

**Comment by nikivazou** [comment]
Thanks @spinda I added your patch deleted all these traces! 

Yes, *eventually* most logical expressions will be written using haskell.

**Comment by ranjitjhala** [comment]
@nikivazou when you say eventually do you mean some feature is missing or is it a matter of just porting the existing aliases?

**Comment by nikivazou** [comment]
I mean porting the existing aliases. 

The only missing feature is maybe type aliases but you had some ideas on how to encode them.

**Comment by ranjitjhala** [comment]
ok, great! so in the tutorial I will then port all the 'predicate' to just
'inline', right?

On Thu, Feb 5, 2015 at 5:49 PM, Niki Vazou <notifications@github.com> wrote:

> I mean porting the existing aliases.
>
> The only missing feature is maybe type aliases but you had some ideas on
> how to encode them.
>
> --
> Reply to this email directly or view it on GitHub
> <https://github.com/ucsd-progsys/liquidhaskell/pull/325#issuecomment-73167393>
> .
>



-- 
Ranjit.

**Comment by nikivazou** [comment]
'predicate' and 'expression'

So if your Haskell function returns a `Bool`, then `inline` is treated as `predicate`, otherwise as `expression`.

**Comment by ranjitjhala** [comment]
nice!

On Thu, Feb 5, 2015 at 5:53 PM, Niki Vazou <notifications@github.com> wrote:

> 'predicate' and 'expression'
>
> So if your Haskell function returns a Bool, then inline is treated as
> predicate, otherwise as expression.
>
> --
> Reply to this email directly or view it on GitHub
> <https://github.com/ucsd-progsys/liquidhaskell/pull/325#issuecomment-73167815>
> .
>



-- 
Ranjit.

**Pull request merged**